shotover/src/codec/cassandra.rs

```diff
@@ -561,46 +659,107 @@ impl CassandraEncoder {
+                        let payload_len = &payload_len.to_le_bytes()[..3];
+                        dst[header_start..header_start + 3].copy_from_slice(payload_len);
+                        dst[header_start + 3..header_start + 6].copy_from_slice(
+                            &cassandra_protocol::crc::crc24(payload_len).to_le_bytes()[..3],
+                        );
 
-                // add payload crc
-                dst.extend_from_slice(
-                    &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
-                );
+                        // add payload crc
+                        dst.extend_from_slice(
+                            &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
+                        );
+                    }
+                    Compression::Lz4 => {
+                        let header_start = dst.len();
+
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
+
+                        // TODO we should not be encoding small frames
+                        let uncompressed_len = self.encode_compressed_payload(dst, m)?;
+                        let compressed_len = dst.len() - payload_start;
+
+                        let mut header =
+                            (compressed_len) as u64 | ((uncompressed_len as u64) << 17);
+
+                        if true {
+                            // TODO if self_contained
+                            header |= 1 << 34;
+                        }
+
+                        let crc = crc24(&header.to_le_bytes()[..5]) as u64;
+
+                        let header = header | (crc << 40);
+
+                        dst[header_start..header_start + 8].copy_from_slice(&header.to_le_bytes());
+
+                        // add payload crc
+                        dst.extend_from_slice(
+                            &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
+                        );
+                    }
+                    _ => unimplemented!("Only Lz4 compression is supported for v5"),
+                }
 
                 Ok(())
             }
-            (_, _) => self.encode_envelope(dst, m),
+            (_, _) => {
+                let message_compression = m.codec_state.as_cassandra();
+                self.encode_envelope(dst, m, message_compression)
+            }
         }
     }
 
-    fn encode_envelope(&mut self, dst: &mut BytesMut, m: Message) -> Result<()> {
-        let message_compression = m.codec_state.as_cassandra();
+    fn encode_compressed_payload(&mut self, dst: &mut BytesMut, m: Message) -> Result<usize> {
+        // TODO: always check if cassandra message
+        Ok(match m.into_encodable(MessageType::Cassandra)? {
+            Encodable::Bytes(bytes) => {
+                let uncompressed_len = bytes.len();
+                let compressed = compress(&bytes);
```

Comment:
https://docs.rs/lz4_flex/latest/lz4_flex/fn.compress_into.html we should use compress_into to directly compress into dst, that will avoid the extra allocation + copy.